 
# Pets Pooled Analysis

## Introduction

This repository contains the code and supplementary material for the companion animal pooled analysis study presented in the manuscript "A comprehensive profile of the companion animal gut microbiome integrating reference-based and reference-free methods" by Branck et al.

Tobyn Branck<sub>1,2,3</sub>, Zhiji Hu1,2, William A. Nickols1,2,4, Aaron M. Walsh1,2,4, Amrisha Bhosle1,2,4, Meghan I. Short1,2,4, Jacob T. Nearing1,2,4, Francesco Asnicar5, Lauren J. McIver1,2, Sagun Maharjan1,2,4, Ali Rahnavard6, Artemis Louyakis3, Dayakar V. Badri3, Christoph Brockel3, Kelsey N. Thompson1,2,4, Curtis Huttenhower1,2,4,7

1Department of Biostatistics, Harvard T.H. Chan School of Public Health, Boston, MA, USA, 2Harvard Chan Microbiome in Public Health Center, Harvard T. H. Chan School of Public Health, Boston, MA, USA, 3Science and Technology Center, Hillâ€™s Pet Nutrition, Inc., Topeka, KS, USA, 4Infectious Disease and Microbiome Program, Broad Institute of MIT and Harvard, Cambridge, MA, USA, 5CIBIO Department, University of Trento, Trento, Italy, 6Computational Biology Institute, Department of Biostatistics and Bioinformatics, Milken Institute School of Public Health, The George Washington University, Washington, DC, USA, 7Department of Immunology and Infectious Diseases, Harvard T. H. Chan School of Public Health, Boston, MA, USA

## Contents
Custom scripts:
* See "scripts" directory for the code used to generate all main and supplemental figures in the manuscript. A detailed README.txt is provided which contains the input files and output files/figures generated by each script.

Supplemental Tables

Includes tables describing individual study characteristics and sample information.
* Supplemental Table 1: 	Public and private data sources of cat, dog, and human gut metagenomic samples.								
* Supplemental Table 2: 	By-sample taxonomic profiles generated by MetaPhlAn 4 for all companion animal gut metagenomes								
* Supplemental Table 3: 	By-sample taxonomic profiles generated by MetaPhlAn 4 for HMP1-II baseline (time 0 samples) gut metagenomic samples 																	
* Supplemental Table 4: 	By-sample taxonomic profiles generated by MetaPhlAn 4 for Madagascar cohort gut metagenomes								
* Supplemental Table 5: 	SGB prevalence by host					
* Supplemental Table 6:	Helicobacteraceae spp. distribution across hosts. 							
* Supplemental Table 7: Campylobacter spp. distributed across hosts.								
* Supplemental Table 8: Feature-wise testing (multivariate linear model) for differences in dog gut microbiome, at the SGB-level, based on housing (facility vs. household).
* Supplemental Table 9: Feature-wise testing (multivariate linear model) for differences in gut microbiome, at the SGB-level, across host species (cats, dogs, and humans).								
* Supplemental Table 10: Feature-wise testing for differences in ARGs across host species (cats, dogs, and humans). Multivariate linear models ran with MaAsLin 2 for all samples.
* Supplemental Table 11: Feature-wise testing for differences in ARGs (organized into antibiotic superclasses) across host species (cats, dogs, and humans). Multivariate linear models ran with MaAsLin 2.
* Supplemental Table 12: Oct22 to Jun23 database taxonomy mapping for SGBs
* Supplemental Table 13: ARG abundance table
* Supplemental Table 14: Abundance table for ARGs grouped into ABX targets (ABX superclasses)
