 
# Pets Pooled Analysis

## Introduction

This repository contains the code and supplementary material for the companion animal pooled analysis study presented in the following manuscript: 

"A comprehensive profile of the companion animal gut microbiome integrating reference-based and reference-free methods"

Tobyn Branck<sup>1,2,3</sup>, Zhiji Hu<sup>1,2</sup>, William A. Nickols<sup>1,2,4</sup>, Aaron M. Walsh<sup>1,2,4</sup>, Amrisha Bhosle<sup>1,2,4</sup>, Meghan I. Short<sup>1,2,4</sup>, Jacob T. Nearing<sup>1,2,4</sup>, Francesco Asnicar<sup>5</sup>, Lauren J. McIver<sup>1,2</sup>, Sagun Maharjan<sup>1,2,4</sup>, Ali Rahnavard<sup>6</sup>, Artemis Louyakis<sup>3</sup>, Dayakar V. Badri<sup>3</sup>, Christoph Brockel<sup>3</sup>, Kelsey N. Thompson<sup>1,2,4</sup>, Curtis Huttenhower<sup>1,2,4,7</sup>
###### <sup>1</sup>Department of Biostatistics, Harvard T.H. Chan School of Public Health, Boston, MA, USA, <sup>2</sup>Harvard Chan Microbiome in Public Health Center, Harvard T. H. Chan School of Public Health, Boston, MA, USA, <sup>3</sup>Science and Technology Center, Hillâ€™s Pet Nutrition, Inc., Topeka, KS, USA, <sup>4</sup>Infectious Disease and Microbiome Program, Broad Institute of MIT and Harvard, Cambridge, MA, USA, <sup>5</sup>CIBIO Department, University of Trento, Trento, Italy, <sup>6</sup>Computational Biology Institute, Department of Biostatistics and Bioinformatics, Milken Institute School of Public Health, The George Washington University, Washington, DC, USA, <sup>7</sup>Department of Immunology and Infectious Diseases, Harvard T. H. Chan School of Public Health, Boston, MA, USA

## Contents
#### Custom scripts:
* See "scripts" directory for the code used to generate all main and supplemental figures in the manuscript. A detailed README.txt is provided which contains the input files and output files/figures generated by each script.

#### Supplemental Tables

Includes tables describing individual study characteristics and sample information.
* Supplemental Table 1: 	Public and private data sources of cat, dog, and human gut metagenomic samples.								
* Supplemental Table 2: 	By-sample taxonomic profiles generated by MetaPhlAn 4 for all companion animal gut metagenomes								
* Supplemental Table 3: 	By-sample taxonomic profiles generated by MetaPhlAn 4 for HMP1-II baseline (time 0 samples) gut metagenomic samples 																	
* Supplemental Table 4: 	By-sample taxonomic profiles generated by MetaPhlAn 4 for Madagascar cohort gut metagenomes								
* Supplemental Table 5: 	SGB prevalence by host					
* Supplemental Table 6:	Helicobacteraceae spp. distribution across hosts. 							
* Supplemental Table 7: Campylobacter spp. distributed across hosts.								
* Supplemental Table 8: Feature-wise testing (multivariate linear model) for differences in dog gut microbiome, at the SGB-level, based on housing (facility vs. household).
* Supplemental Table 9: Feature-wise testing (multivariate linear model) for differences in gut microbiome, at the SGB-level, across host species (cats, dogs, and humans).								
* Supplemental Table 10: Feature-wise testing for differences in ARGs across host species (cats, dogs, and humans). Multivariate linear models ran with MaAsLin 2 for all samples.
* Supplemental Table 11: Feature-wise testing for differences in ARGs (organized into antibiotic superclasses) across host species (cats, dogs, and humans). Multivariate linear models ran with MaAsLin 2.
* Supplemental Table 12: Oct22 to Jun23 database taxonomy mapping for SGBs
* Supplemental Table 13: ARG abundance table
* Supplemental Table 14: Abundance table for ARGs grouped into ABX targets (ABX superclasses)
* Supplemental Table 15: sample-specific metadata
